sdv/_utils.py

```diff
@@ -200,3 +201,80 @@ def _format_invalid_values_string(invalid_values, num_values):
             return f'{invalid_values[:num_values] + extra_missing_values}'
 
     return f'{invalid_values}'
+
+
+def _find_root_tables(relationships):
+    parent_tables = {rel['parent_table_name'] for rel in relationships}
+    child_tables = {rel['child_table_name'] for rel in relationships}
+    return parent_tables - child_tables
+
+
+def _get_relationship_for_child(relationships, child_table):
+    return [rel for rel in relationships if rel['child_table_name'] == child_table]
+
+
+def _get_relationship_idx_for_parent(relationships, parent_table):
+    return [
+        idx for idx, rel in enumerate(relationships) if rel['parent_table_name'] == parent_table
+    ]
+
+
+def _get_rows_to_drop(metadata, data):
+    """Get the rows to drop to ensure referential integrity.
+
+    The logic of this function is to start at the root tables, look at invalid references
+    and then save the index of the rows to drop. Then, we looked at the relationships that
+    we didn't check and repeat the process until there are no more relationships to check.
+    This ensures that we preserve the referential integrity between all the relationships.
+
+    Args:
+        metadata (MultiTableMetadata):
+            Metadata of the datasets.
+        data (dict):
+            Dictionary that maps each table name (string) to the data for that
+            table (pandas.DataFrame).
+
+    Returns:
+        dict:
+            Dictionary with the table names as keys and the indexes of the rows to drop as values.
+    """
+    table_to_idx_to_drop = {}
+    relationships = deepcopy(metadata.relationships)
+    while relationships:
+        current_roots = _find_root_tables(relationships)
+        for root in current_roots:
+            relationship_idx = _get_relationship_idx_for_parent(relationships, root)
+            for idx in relationship_idx:
+                relationship = relationships[idx]
+                parent_table = relationship['parent_table_name']
+                child_table = relationship['child_table_name']
+                parent_column = relationship['parent_primary_key']
+                child_column = relationship['child_foreign_key']
+                if child_table not in table_to_idx_to_drop:
+                    table_to_idx_to_drop[child_table] = set()
+
+                is_nan = data[child_table][child_column].isna()
+                valid_parent_idx = [
+                    idx for idx in data[parent_table].index if idx not in table_to_idx_to_drop.get(
+                        parent_table, set()
+                    )
+                ]
+                valid_parent_values = set(data[parent_table].loc[valid_parent_idx, parent_column])
```

Comment:
Can we move this to doing it once outside of the for loop as well? Once we're evaluating it as a root table, the valid parent values shouldn't change again, right?

Reply:
Yes, done in [d1777fc](https://github.com/sdv-dev/SDV/pull/1800/commits/d1777fc82ca5ea78d296810b1e7fd8398127a689)